README.rst
==============
k2hdkc_python
==============

.. image:: https://img.shields.io/badge/license-MIT-blue.svg
        :target: https://github.com/yahoojapan/k2hdkc_python/blob/master/LICENSE
.. image:: https://img.shields.io/pypi/pyversions/k2hdkc.svg
        :target: https://pypi.python.org/pypi/k2hdkc
.. image:: https://img.shields.io/github/forks/yahoojapan/k2hdkc_python.svg
        :target: https://github.com/yahoojapan/k2hdkc_python/network
.. image:: https://img.shields.io/github/stars/yahoojapan/k2hdkc_python.svg
        :target: https://github.com/yahoojapan/k2hdkc_python/stargazers
.. image:: https://img.shields.io/github/issues/yahoojapan/k2hdkc_python.svg
        :target: https://github.com/yahoojapan/k2hdkc_python/issues
.. image:: https://github.com/yahoojapan/k2hdkc_python/workflows/Python%20package/badge.svg
        :target: https://github.com/yahoojapan/k2hdkc_python/actions
.. image:: https://readthedocs.org/projects/k2hdkc-python/badge/?version=latest
        :target: https://k2hdkc-python.readthedocs.io/en/latest/?badge=latest
.. image:: https://img.shields.io/pypi/v/k2hdkc
        :target: https://pypi.org/project/k2hdkc/



Overview
---------

k2hdkc_python is an official python driver for `k2hdkc`_.

.. _`k2hdkc`: https://k2hdkc.antpick.ax/

.. image:: https://raw.githubusercontent.com/yahoojapan/k2hdkc_python/main/docs/images/top_k2hdkc_python.png


Install
--------

Let's install k2hdkc using pip::

    pip install k2hdkc


Usage
------

Firstly you must install the k2hdkc shared library::

    $ curl -o- https://raw.github.com/yahoojapan/k2hdkc_python/master/cluster/start_server.sh | bash


Then, Let's try to set a key and get it::

    import k2hdkc
    
    k = k2hdkc.K2hdkc('slave.yaml')
    k.set('hello', 'world')
    v = k.get('hello')
    print(v)    // world


Development
------------

Clone this repository and go into the directory, then run the following command::

    $ python3 -m pip install --upgrade build
    $ python3 -m build


Documents
----------

Here are documents including other components.

`Document top page`_

`About K2HDKC`_

`About AntPickax`_

.. _`Document top page`: https://k2hdkc-python.readthedocs.io/
.. _`ドキュメントトップ`: https://k2hdkc-python.readthedocs.io/
.. _`About K2HDKC`: https://k2hdkc.antpick.ax/
.. _`K2HDKCについて`: https://k2hdkc.antpick.ax/
.. _`About AntPickax`: https://antpick.ax
.. _`AntPickaxについて`: https://antpick.ax


Packages
--------

Here are packages including other components.

`k2hdkc(python packages)`_

.. _`k2hdkc(python packages)`:  https://pypi.org/project/k2hdkc/


License
--------

MIT License. See the LICENSE file.

AntPickax
---------

**k2hdkc_python** is a project by AntPickax_, which is an open source team in `Yahoo Japan Corporation`_.

.. _AntPickax: https://antpick.ax/
.. _`Yahoo Japan Corporation`: https://about.yahoo.co.jp/info/en/company/



HISTORY.rst
=======
History
=======

1.0.1 (2023-10-02)
-------------------

* Updates the CI settings

1.0.0 (2022-02-07)
-------------------

* First release on PyPI.


src/k2hdkc/tests/test_k2hdkc.py
# -*- coding: utf-8 -*-
#
# K2hdkc Python Driver under MIT License
#
# Copyright (c) 2022 Yahoo Japan Corporation
#
# For the full copyright and license information, please view
# the license file that was distributed with this source code.
#
# AUTHOR:   Hirotaka Wakabayashi
# CREATE:   Tue Feb 08 2022
# REVISION:
#

import ctypes
import logging
import time
import unittest
from struct import pack

import k2hdkc


class TestK2hdkc(unittest.TestCase):
    def test_K2hdkc_construct(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        db.close()

    def test_K2hdkc_add_subkey(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        subkey = "sub_hello"
        subval = "sub_world"
        self.assertTrue(db.add_subkey(key, subkey, subval), True)
        self.assertTrue(db.get_subkeys(key), [subkey])
        db.close()

    def test_K2hdkc_clear_subkeys(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        subkey = "sub_hello"
        subval = "sub_world"
        self.assertTrue(db.add_subkey(key, subkey, subval), True)
        self.assertTrue(db.get_subkeys(key), [subkey])
        self.assertTrue(db.clear_subkeys(key), True)
        # TODO error
        # self.assertTrue(db.get_subkeys(key), [])
        db.close()

    def test_K2hdkc_get(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        db.close()

    @unittest.skip("skipping because no attrs")
    def test_K2hdkc_get_attrs(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        db.close()

    def test_K2hdkc_get_subkeys(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        subkey = "sub_hello"
        subval = "sub_world"
        self.assertTrue(db.add_subkey(key, subkey, subval), True)
        self.assertTrue(db.get_subkeys(key), [subkey])
        db.close()

    def test_K2hdkc_queue_get(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        qkey = "q_hello"
        qval = "q_world"
        self.assertTrue(db.queue_put(qkey, qval), True)
        self.assertTrue(db.queue_get(qkey), qval)
        db.close()

    def test_K2hdkc_queue_put(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        qkey = "q_hello"
        qval = "q_world"
        self.assertTrue(db.queue_put(qkey, qval), True)
        self.assertTrue(db.queue_get(qkey), qval)
        db.close()

    def test_K2hdkc_keyqueue_get(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        qprefix = "q_prefix"
        qkey = "q_hello"
        qval = "q_world"
        self.assertTrue(db.keyqueue_put(qprefix, qkey, qval), True)
        self.assertTrue(db.keyqueue_get(qprefix), {qkey, qval})
        db.close()

    def test_K2hdkc_keyqueue_put(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        qprefix = "q_prefix"
        qkey = "q_hello"
        qval = "q_world"
        self.assertTrue(db.keyqueue_put(qprefix, qkey, qval), True)
        self.assertTrue(db.keyqueue_get(qprefix), {qkey, qval})
        db.close()

    def test_K2hdkc_remove(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        self.assertTrue(db.remove(key), True)
        self.assertFalse(db.get(key), val)
        db.close()

    def test_K2hdkc_remove_subkeys(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        subkey = "sub_hello"
        subval = "sub_world"
        self.assertTrue(db.add_subkey(key, subkey, subval), True)
        self.assertTrue(db.get_subkeys(key), [subkey])
        # TODO error
        # self.assertTrue(db.remove_subkeys(key, subkey), True)
        db.close()

    def test_K2hdkc_rename(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        newkey = "olleh"
        self.assertTrue(db.rename(key, newkey), True)
        self.assertTrue(db.get(newkey), val)
        db.close()

    def test_K2hdkc_set(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        db.close()

    def test_K2hdkc_set_subkeys(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        subkey = "sub_hello"
        subval = "sub_world"
        self.assertTrue(db.add_subkey(key, subkey, subval), True)
        self.assertTrue(db.get_subkeys(key), [subkey])
        db.close()

    def test_K2hdkc_cas_init(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        cas_key = "cas_hello"
        cas_value = 65530
        self.assertTrue(db.cas_init(cas_key, pack("H", cas_value)), True)
        self.assertTrue(db.cas_get(cas_key, k2hdkc.DataType.U_SHORT), cas_value)
        db.close()

    def test_K2hdkc_cas_get(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        cas_key = "cas_hello"
        cas_value = 65530
        self.assertTrue(db.cas_init(cas_key, pack("H", cas_value)), True)
        self.assertTrue(db.cas_get(cas_key, k2hdkc.DataType.U_SHORT), cas_value)
        db.close()

    def test_K2hdkc_cas_decrement(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        cas_key = "cas_hello"
        cas_value = 65530
        self.assertTrue(db.cas_init(cas_key, pack("H", cas_value)), True)
        self.assertTrue(db.cas_get(cas_key, k2hdkc.DataType.U_SHORT), cas_value)
        self.assertTrue(db.cas_increment(cas_key), True)
        self.assertTrue(db.cas_get(cas_key, k2hdkc.DataType.U_SHORT), cas_value + 1)
        self.assertTrue(db.cas_decrement(cas_key), True)
        self.assertTrue(db.cas_get(cas_key, k2hdkc.DataType.U_SHORT), cas_value)
        db.close()

    def test_K2hdkc_cas_increment(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        cas_key = "cas_hello"
        cas_value = 65530
        self.assertTrue(db.cas_init(cas_key, pack("H", cas_value)), True)
        self.assertTrue(db.cas_get(cas_key, k2hdkc.DataType.U_SHORT), cas_value)
        self.assertTrue(db.cas_increment(cas_key), True)
        self.assertTrue(db.cas_get(cas_key, k2hdkc.DataType.U_SHORT), cas_value + 1)
        self.assertTrue(db.cas_decrement(cas_key), True)
        self.assertTrue(db.cas_get(cas_key, k2hdkc.DataType.U_SHORT), cas_value)
        db.close()

    def test_K2hdkc_cas_set(self):
        db = k2hdkc.K2hdkc("/tmp/slave.yaml")
        self.assertTrue(isinstance(db, k2hdkc.K2hdkc))
        key = "hello"
        val = "world"
        self.assertTrue(db.set(key, val), True)
        self.assertTrue(db.get(key), val)
        cas_key = "cas_hello"
        cas_value = 65530
        self.assertTrue(db.cas_init(cas_key, pack("H", cas_value)), True)
        self.assertTrue(db.cas_get(cas_key, k2hdkc.DataType.U_SHORT), cas_value)
        self.assertTrue(db.cas_increment(cas_key), True)
        self.assertTrue(db.cas_get(cas_key, k2hdkc.DataType.U_SHORT), cas_value + 1)
        self.assertTrue(db.cas_decrement(cas_key), True)
        self.assertTrue(db.cas_get(cas_key, k2hdkc.DataType.U_SHORT), cas_value)
        new_cas_value = 1
        self.assertTrue(
            db.cas_set(cas_key, pack("H", cas_value), pack("H", new_cas_value)), True
        )
        self.assertTrue(db.cas_get(cas_key, k2hdkc.DataType.U_SHORT), new_cas_value)
        db.close()


if __name__ == "__main__":
    unittest.main()

#
# Local variables:
# tab-width: 4
# c-basic-offset: 4
# End:
# vim600: expandtab sw=4 ts=4 fdm=marker
# vim<600: expandtab sw=4 ts=4
#


.github/scripts/test_with_unittest.sh
#!/bin/sh
#
# -*- coding: utf-8 -*-
#
# K2hdkc Python Driver under MIT License
#
# Copyright (c) 2022 Yahoo Japan Corporation
#
# For the full copyright and license information, please view
# the license file that was distributed with this source code.
#
# AUTHOR:   Hirotaka Wakabayashi
# CREATE:   Tue Feb 08 2022
# REVISION:
#

echo $(basename $0)

if test -f "/etc/os-release"; then
    . /etc/os-release
    OS_NAME=$ID
    OS_VERSION=$VERSION_ID
elif test -f "/etc/centos-release"; then
    echo "[OK] /etc/centos-release falling back to CentOS-7"
    OS_NAME=centos
    OS_VERSION=7
else
    echo "[NO] Unknown OS, neither /etc/os-release nor /etc/centos-release"
    exit 1
fi

echo "[OK] HOSTNAME=${HOSTNAME} OS_NAME=${OS_NAME} OS_VERSION=${OS_VERSION}"
PYTHON=""
case "${OS_NAME}-${OS_VERSION}" in
    ubuntu*|debian*)
        PYTHON=$(which python)
        ;;
    centos*|fedora*)
        PYTHON=$(which python3)
        ;;
esac

cd src

TEST_FILES="test_k2hdkc.py test_k2hdkc_package.py"
for TEST_FILE in ${TEST_FILES}
do
    ${PYTHON} -m unittest k2hdkc/tests/${TEST_FILE}
    if test $? -ne 0; then
        echo "[NO] ${PYTHON} -m unittest k2hdkc/tests/${TEST_FILE}"
        exit 1
    fi
done

exit 0

#
# Local variables:
# tab-width: 4
# c-basic-offset: 4
# End:
# vim600: expandtab sw=4 ts=4 fdm=marker
# vim<600: expandtab sw=4 ts=4
#


src/k2hdkc/tests/test_k2hdkc_package.py
# -*- coding: utf-8 -*-
#
# K2hdkc Python Driver under MIT License
#
# Copyright (c) 2022 Yahoo Japan Corporation
#
# For the full copyright and license information, please view
# the license file that was distributed with this source code.
#
# AUTHOR:   Hirotaka Wakabayashi
# CREATE:   Tue Feb 08 2022
# REVISION:
#

import logging
import unittest

import k2hdkc


class TestK2hdkcPackage(unittest.TestCase):
    def test_get_library_handle(self):
        libk2hdkc = k2hdkc.get_library_handle()
        self.assertTrue(libk2hdkc)
        self.assertTrue(isinstance(libk2hdkc, dict))
        self.assertTrue(libk2hdkc["c"])
        self.assertTrue(libk2hdkc["k2hdkc"])

    def test_set_log_level(self):
        k2hdkc.set_log_level(logging.INFO)
        logger = logging.getLogger("k2hdkc")
        self.assertEqual(logging.getLevelName(logger.level), "INFO")

    def test_set_layer_log_level(self):
        self.assertEqual(k2hdkc.set_layer_log_level(k2hdkc.LayerLogLevel.K2HDKC), True)


if __name__ == "__main__":
    unittest.main()

#
# Local variables:
# tab-width: 4
# c-basic-offset: 4
# End:
# vim600: expandtab sw=4 ts=4 fdm=marker
# vim<600: expandtab sw=4 ts=4
#


cluster/k2hdkc.sh
#!/bin/sh
#
# -*- coding: utf-8 -*-
#
# K2hdkc Python Driver under MIT License
#
# Copyright (c) 2022 Yahoo Japan Corporation
#
# For the full copyright and license information, please view
# the license file that was distributed with this source code.
#
# AUTHOR:   Hirotaka Wakabayashi
# CREATE:   Tue Feb 08 2022
# REVISION:
#
#

# A K2hdkc Cluster management script.

USER=$(whoami)
BASEDIR=$(dirname "$0")

start_process() {
	echo "chmpx -conf ${BASEDIR}/server.yaml -d dump > ${BASEDIR}/chmpx.log 2>&1"
	nohup chmpx -conf ${BASEDIR}/server.yaml -d dump > ${BASEDIR}/chmpx.log 2>&1 &
	echo "sleep 3"
	sleep 3
	echo "k2hdkc -conf ${BASEDIR}/server.yaml -d dump > ${BASEDIR}/k2hdkc.log 2>&1"
	nohup k2hdkc -conf ${BASEDIR}/server.yaml -d dump > ${BASEDIR}/k2hdkc.log 2>&1 &
	echo "sleep 3"
	sleep 3
	echo "chmpx -conf ${BASEDIR}/slave.yaml -d dump  > ${BASEDIR}/slave.log 2>&1"
	nohup chmpx -conf ${BASEDIR}/slave.yaml -d dump  > ${BASEDIR}/slave.log 2>&1 &
	echo "sleep 3"
	sleep 3
}

stop_process() {
	for PROC in chmpx k2hdkc; do
		echo "pgrep -u ${USER} -x ${PROC}"
		pgrep -u ${USER} -x ${PROC}
		if test "${?}" = "0"; then
			echo "pkill -x ${PROC}"
			pkill -x ${PROC}
		fi
	done
}

status_process() {
	RET=0
	for PROC in chmpx k2hdkc; do
		echo "pgrep -u ${USER} -x ${PROC}"
		pgrep -u ${USER} -x ${PROC}
		RET=$(echo ${?} + ${RET}|bc)
	done
	return ${RET}
}

case $1 in
startifnotexist)
	status_process
	if test "${RET}" = "2" ; then
		echo "start_process"
		start_process
	elif test "${RET}" = "1" ; then
		echo "stop_process"
		stop_process
		echo "start_process"
		start_process
	fi
	;;
start)
	start_process
	;;
stop)
	stop_process
	;;
status)
	status_process
	;;
*)
	echo "$0 (start|stop|status|startifnotexist)"
	exit 1
	;;
esac
exit 0

#
# Local variables:
# tab-width: 4
# c-basic-offset: 4
# End:
# vim600: expandtab sw=4 ts=4 fdm=marker
# vim<600: expandtab sw=4 ts=4
#



.github/scripts/lint_with_pylint.sh
#!/bin/sh
#
# -*- coding: utf-8 -*-
#
# K2hdkc Python Driver under MIT License
#
# Copyright (c) 2022 Yahoo Japan Corporation
#
# For the full copyright and license information, please view
# the license file that was distributed with this source code.
#
# AUTHOR:   Hirotaka Wakabayashi
# CREATE:   Tue Feb 08 2022
# REVISION:
#

echo $(basename $0)

if test -f "/etc/os-release"; then
    . /etc/os-release
    OS_NAME=$ID
    OS_VERSION=$VERSION_ID
elif test -f "/etc/centos-release"; then
    echo "[OK] /etc/centos-release falling back to CentOS-7"
    OS_NAME=centos
    OS_VERSION=7
else
    echo "Unknown OS, neither /etc/os-release nor /etc/centos-release"
    exit 1
fi

echo "[OK] HOSTNAME=${HOSTNAME} OS_NAME=${OS_NAME} OS_VERSION=${OS_VERSION}"
pylint src/k2hdkc

exit $?

#
# Local variables:
# tab-width: 4
# c-basic-offset: 4
# End:
# vim600: expandtab sw=4 ts=4 fdm=marker
# vim<600: expandtab sw=4 ts=4
#


cluster/start_server.sh
#!/bin/sh
#
# -*- coding: utf-8 -*-
#
# K2hdkc Python Driver under MIT License
#
# Copyright (c) 2022 Yahoo Japan Corporation
#
# For the full copyright and license information, please view
# the license file that was distributed with this source code.
#
# AUTHOR:   Hirotaka Wakabayashi
# CREATE:   Tue Feb 08 2022
# REVISION:
#

# Sets PATH. setup_*.sh uses useradd command
PATH=${PATH}:/usr/sbin:/sbin

# an unset parameter expansion will fail
set -u

# umask 022 is enough
umask 022

# environments
SRCDIR=$(cd $(dirname "$0") && pwd)
DEBUG=1
if test "${DEBUG}" -eq 1; then
	TAG="$(basename $0) -s"
else
	TAG=$(basename $0)
fi
USER=$(whoami)
LOGLEVEL=info

# Starts chmpx and k2hdkc
#
start_process() {
	if ! test -f "/tmp/server.yaml"; then
		echo "cp ${SRCDIR}/server.yaml /tmp/"
		cp ${SRCDIR}/server.yaml /tmp
	fi
	if ! test -f "/tmp/slave.yaml"; then
		echo "cp ${SRCDIR}/slave.yaml /tmp/"
		cp ${SRCDIR}/slave.yaml /tmp
	fi

	echo "chmpx -conf /tmp/server.yaml -d ${LOGLEVEL} > /tmp/chmpx.log 2>&1"
	nohup chmpx -conf /tmp/server.yaml -d ${LOGLEVEL} > /tmp/chmpx.log 2>&1 &
	echo "sleep 3"
	sleep 3
	echo "k2hdkc -conf /tmp/server.yaml -d ${LOGLEVEL} > /tmp/k2hdkc.log 2>&1"
	nohup k2hdkc -conf /tmp/server.yaml -d ${LOGLEVEL} > /tmp/k2hdkc.log 2>&1 &
	echo "sleep 3"
	sleep 3
	echo "chmpx -conf /tmp/slave.yaml -d ${LOGLEVEL}  > /tmp/slave.log 2>&1"
	nohup chmpx -conf /tmp/slave.yaml -d ${LOGLEVEL}  > /tmp/slave.log 2>&1 &
	echo "sleep 3"
	sleep 3
}

# Stops chmpx and k2hdkc
#
stop_process() {
	for PROC in chmpx k2hdkc; do
		echo "pgrep -u ${USER} -x ${PROC}"
		pgrep -u ${USER} -x ${PROC}
		if test "${?}" = "0"; then
			echo "pkill -9 -x ${PROC}"
			pkill -9 -x ${PROC}
		fi
	done
	if test -f "/tmp/server.yaml"; then
		echo "rm -f /tmp/server.yaml"
		rm -f /tmp/server.yaml
	fi
	if test -f "/tmp/slave.yaml"; then
		echo "rm -f /tmp/slave.yaml"
		rm -f /tmp/slave.yaml
	fi
}

# Shows status of chmpx and k2hdkc
#
status_process() {
	RET=0
	for PROC in chmpx k2hdkc; do
		echo "pgrep -u ${USER} -x ${PROC}"
		pgrep -u ${USER} -x ${PROC}
		RET=$(echo ${?} + ${RET}|bc)
	done
	return ${RET}
}

# Starts chmpx and k2hdkc if no chmpx and k2hdkc processes are running
#
startifnotexist() {
	status_process
	if test "${RET}" = "2" ; then
		echo "start_process"
		start_process
	elif test "${RET}" = "1" ; then
		echo "stop_process"
		stop_process
		echo "start_process"
		start_process
	fi
}

# Checks if k2hdkc is installed 
#
# Params::
#   no params
#
# Returns::
#   0 on installed
#   1 on not installed
#
which_k2hdkc() {
	which k2hdkc
	if test "${?}" = "0"; then
		echo "[OK] k2hdkc already installed"
		return 0
	fi
	return 1
}

# Determines the current OS
#
# Params::
#   no params
#
# Returns::
#   0 on success
#   1 on failure
#
setup_os_env() {
	if test -f "/etc/os-release"; then
		. /etc/os-release
		OS_NAME=$ID
		OS_VERSION=$VERSION_ID
	else
		echo "[OK] unknown OS, no /etc/os-release and /etc/centos-release falling back to CentOS-7"
		OS_NAME=centos
		OS_VERSION=7
	fi

	if test "${OS_NAME}" = "ubuntu"; then
		echo "[OK] ubuntu configurations are currently equal to debian one"
		OS_NAME=debian
	fi

	HOSTNAME=$(hostname)
	echo "[OK] HOSTNAME=${HOSTNAME} OS_NAME=${OS_NAME} OS_VERSION=${OS_VERSION}"
}

# Builds k2hash from source code
#
# Params::
#   $1 os_name
#
# Returns::
#   0 on success
#   1 on failure(exit)
#
make_k2hash() {

	_os_name=${1:?"os_name should be nonzero"}

	if test "${_os_name}" = "debian" -o "${_os_name}" = "ubuntu"; then
		_configure_opt="--with-gcrypt"
		sudo apt-get update -y
		sudo apt-get install -y git curl autoconf autotools-dev gcc g++ make gdb libtool pkg-config libyaml-dev libgcrypt20-dev
	elif test "${_os_name}" = "fedora"; then
		_configure_opt="--with-nss"
		sudo dnf install -y git curl autoconf automake gcc gcc-c++ gdb make libtool pkgconfig libyaml-devel nss-devel
	elif test "${_os_name}" = "centos" -o "${_os_name}" = "rhel"; then
		_configure_opt="--with-nss"
		if test "${OS_VERSION}" = "7"; then
			sudo yum install -y git curl autoconf automake gcc gcc-c++ gdb make libtool pkgconfig libyaml-devel nss-devel
		elif test "${OS_VERSION}" = "8"; then
			sudo dnf install -y git curl autoconf automake gcc gcc-c++ gdb make libtool pkgconfig
			sudo dnf install -y --enablerepo=powertools nss-devel libyaml-devel
		fi
	else
		echo "[NO] OS should be debian, ubuntu, fedora, centos or rhel"
		exit 1
	fi

	echo "[OK] git clone https://github.com/yahoojapan/fullock"
	git clone https://github.com/yahoojapan/fullock
	echo "[OK] git clone https://github.com/yahoojapan/k2hash"
	git clone https://github.com/yahoojapan/k2hash
	
	if ! test -d "fullock"; then
		echo "no fullock"
		exit 1
	fi
	cd fullock
	./autogen.sh
	./configure --prefix=/usr
	make
	sudo make install
	
	if ! test -d "../k2hash"; then
		echo "no k2hash"
		exit 1
	fi
	cd ../k2hash
	./autogen.sh
	./configure --prefix=/usr ${_configure_opt}
	make
	sudo make install
	
	return 0
}

# Builds k2hdkc from source code
#
# Params::
#   $1 os_name
#
# Returns::
#   0 on success
#   1 on failure(exit)
#
make_k2hdkc() {

	_os_name=${1:?"os_name should be nonzero"}

	make_k2hash ${_os_name}

	if test "${_os_name}" = "debian" -o "${_os_name}" = "ubuntu"; then
		_configure_opt="--with-gnutls"
		sudo apt-get update -y
		sudo apt-get install -y git curl autoconf autotools-dev gcc g++ make gdb libtool pkg-config libyaml-dev libgnutls28-dev
	elif test "${_os_name}" = "fedora"; then
		_configure_opt="--with-nss"
		sudo dnf install -y git curl autoconf automake gcc gcc-c++ gdb make libtool pkgconfig libyaml-devel nss-devel
	elif test "${_os_name}" = "centos" -o "${_os_name}" = "rhel"; then
		_configure_opt="--with-nss"
		if test "${OS_VERSION}" = "7"; then
			sudo yum install -y git curl autoconf automake gcc gcc-c++ gdb make libtool pkgconfig libyaml-devel nss-devel
		elif test "${OS_VERSION}" = "8"; then
			sudo dnf install -y git curl autoconf automake gcc gcc-c++ gdb make libtool pkgconfig
			sudo dnf install -y --enablerepo=powertools nss-devel libyaml-devel
		fi
	else
		echo "[NO] OS should be debian, ubuntu, fedora, centos or rhel"
		exit 1
	fi

	echo "[OK] git clone https://github.com/yahoojapan/k2hdkc"
	git clone https://github.com/yahoojapan/k2hdkc
	cd k2hdkc
	
	echo "[OK] git clone https://github.com/yahoojapan/chmpx"
	git clone https://github.com/yahoojapan/chmpx
	
	cd chmpx
	./autogen.sh
	./configure --prefix=/usr ${_configure_opt}
	make
	sudo make install
	
	cd ..
	./autogen.sh
	./configure --prefix=/usr ${_configure_opt}
	make
	sudo make install

	which k2hdkc
	if test "${?}" != "0"; then
		echo "[NO] no k2hdkc installed"
		exit 1
	fi
	return 0
}

#
# main loop
#

setup_os_env

which_k2hdkc
if test "${?}" = "0"; then
	startifnotexist
	exit 0
fi

if test "${OS_NAME}" = "fedora"; then
	which sudo
	if test "${?}" = "1"; then
		dnf install -y sudo
	fi
	which bc
	if test "${?}" = "1"; then
		sudo dnf install -y bc
	fi
	sudo dnf install -y curl
	curl -s https://packagecloud.io/install/repositories/antpickax/current/script.rpm.sh | sudo bash
	sudo dnf install  k2hdkc-devel -y
elif test "${OS_NAME}" = "debian" -o "${OS_NAME}" = "ubuntu"; then
	which sudo
	if test "${?}" = "1"; then
		apt-get update -y
		apt-get install -y sudo
	fi
	which bc
	if test "${?}" = "1"; then
		sudo apt-get install -y bc
	fi
	sudo apt-get install -y curl
	curl -s https://packagecloud.io/install/repositories/antpickax/stable/script.deb.sh | sudo bash
	sudo apt-get install -y k2hdkc-dev
elif test "${OS_NAME}" = "centos" -o "${OS_NAME}" = "rhel"; then
	which sudo
	if test "${?}" = "1"; then
		if test "${OS_VERSION}" = "7"; then
			sudo yum install -y sudo
		elif test "${OS_VERSION}" = "8"; then
			sudo dnf install -y sudo
		fi
	fi
	which bc
	if test "${?}" = "1"; then
		if test "${OS_VERSION}" = "7"; then
			sudo yum install -y bc
		elif test "${OS_VERSION}" = "8"; then
			sudo dnf install -y bc
		fi
	fi
	if test "${OS_VERSION}" = "7"; then
		sudo yum install -y k2hdkc-devel
	elif test "${OS_VERSION}" = "8"; then
		sudo dnf install -y k2hdkc-devel
	fi
	if test "${?}" != "0"; then
		if test "${OS_VERSION}" = "7"; then
			sudo yum install -y curl
		elif test "${OS_VERSION}" = "8"; then
			sudo dnf install -y curl
		fi
		curl -s https://packagecloud.io/install/repositories/antpickax/stable/script.rpm.sh | sudo bash
		if test "${OS_VERSION}" = "7"; then
			sudo yum install -y k2hdkc-devel
		elif test "${OS_VERSION}" = "8"; then
			sudo dnf install -y k2hdkc-devel
		fi
	fi
else
	echo "[NO] OS must be either fedora or centos or debian or ubuntu, not ${OS_NAME}"
	exit 1
fi

which_k2hdkc
if test "${?}" = "0"; then
	startifnotexist
	exit 0
else
	echo "[NO] k2hdkc"
	exit 1
fi

exit 0

#
# Local variables:
# tab-width: 4
# c-basic-offset: 4
# End:
# vim600: expandtab sw=4 ts=4 fdm=marker
# vim<600: expandtab sw=4 ts=4
#
